Source/Core/Core/IOS/Network/IP/Top.cpp

```diff
@@ -162,70 +167,232 @@ static s32 MapWiiSockOptNameToNative(u32 optname)
+                                 std::bit_cast<in_addr>(entry.dwForwardMask),
+                                 std::bit_cast<in_addr>(entry.dwForwardNextHop));
+    }
+
+    if (result == NO_ERROR)
+      break;
+
+    result = GetIpForwardTable(forward_table.get(), &forward_table_size, FALSE);
+  }
+
+#elif defined(__linux__)
+  constexpr int BUFF_SIZE = 8192;
+  constexpr timeval socket_timeout{.tv_sec = 2, .tv_usec = 0};
+  unsigned int msg_seq = 0;
+  const int sock = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
+  if (sock < 0)
+  {
+    ERROR_LOG_FMT(IOS_NET, "Failed to open netlink socket, error {}", Common::StrNetworkError());
+    return {};
+  }
+
+  if (setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, static_cast<const void*>(&socket_timeout),
+                 sizeof(socket_timeout)) < 0)
+  {
+    ERROR_LOG_FMT(IOS_NET, "Failed to set netlink socket recv timeout: {}",
+                  Common::StrNetworkError());
+    return {};
+  }
+
+  Common::ScopeGuard socket_guard{[sock] { close(sock); }};
+  std::array<char, BUFF_SIZE> msg_buffer{};
+  auto nl_msg = reinterpret_cast<nlmsghdr*>(msg_buffer.data());
+  auto rt_msg = reinterpret_cast<rtmsg*>(NLMSG_DATA(nl_msg));
+  const unsigned int pid = getpid();
+
+  // prepare command/netlink packet
+  nl_msg->nlmsg_len = NLMSG_LENGTH(sizeof(rtmsg));  // Length of message.
+  nl_msg->nlmsg_type = RTM_GETROUTE;                // Get the routes from kernel routing table .
+
+  nl_msg->nlmsg_flags = NLM_F_DUMP | NLM_F_REQUEST;  // The message is a request for dump.
+  nl_msg->nlmsg_seq = msg_seq++;                     // Sequence of the message packet.
+  nl_msg->nlmsg_pid = pid;                           // PID of process sending the request.
+
+  // ship it
+  if (send(sock, nl_msg, nl_msg->nlmsg_len, 0) < 0)
+  {
+    ERROR_LOG_FMT(IOS_NET, "Failed to send netlink request ({})", Common::StrNetworkError());
+    return {};
+  }
+
+  // read response
+  int msg_len = 0;
+  msg_buffer.fill(0);
+
+  do
+  {
+    auto buf_ptr = msg_buffer.data() + msg_len;
+    const int read_len = recv(sock, buf_ptr, BUFF_SIZE - msg_len, 0);
+    if (read_len < 0)
+    {
+      ERROR_LOG_FMT(IOS_NET, "Failed to receive netlink response ({})", Common::StrNetworkError());
+      return {};
+    }
+
+    nl_msg = reinterpret_cast<nlmsghdr*>(buf_ptr);
+    if (NLMSG_OK(nl_msg, read_len) == 0)
+    {
+      ERROR_LOG_FMT(IOS_NET, "Received netlink error response ({})", NLMSG_OK(nl_msg, read_len));
+      return {};
+    }
+
+    if (nl_msg->nlmsg_type == NLMSG_ERROR)
+    {
+      auto err = reinterpret_cast<nlmsgerr*>(NLMSG_DATA(nl_msg));
+      if (nl_msg->nlmsg_len < NLMSG_LENGTH(sizeof(nlmsgerr)))
       {
-        ifIndex = forward_table->table[i].dwForwardIfIndex;
-        break;
+        ERROR_LOG_FMT(IOS_NET, "Netlink message was truncated");
       }
+      else
+      {
+        ERROR_LOG_FMT(IOS_NET, "Received netlink error {}: {}", err->error,
+                      Common::DecodeNetworkError(err->error));
```

Comment:
```suggestion
        ERROR_LOG_FMT(IOS_NET, "Received netlink error {}: {}", err->error,
                      Common::DecodeNetworkError(-err->error));
```

Reply:
isn't err->error already a negative number? or is this to make it now show up as a negative error?

Reply:
[errno](https://www.man7.org/linux/man-pages/man3/errno.3.html) have positive error values. Some functions might return or store negative errno values _(to indicate an error, i.e. return value < 0)_. That's why in some cases you'll need to change the value sign.

Looking at [github repositories](https://github.com/search?q=strerror+AND+nlmsgerr&type=code) also suggests that the error value needs to be negated.

Reply:
sounds weird to me with the call returning negative (or unsigned) value, but fair enough